incubator/virtualcluster/pkg/syncer/syncer.go

```diff
@@ -17,34 +17,52 @@ limitations under the License.
 package syncer
 
 import (
+	"fmt"
+	"sync"
 	"time"
 
+	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/informers"
 	clientset "k8s.io/client-go/kubernetes"
 	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/tools/clientcmd"
-
+	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
 
 	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/apis/tenancy/v1alpha1"
 	vcinformers "github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	vclisters "github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/client/listers/tenancy/v1alpha1"
 	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/cluster"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/constants"
 	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/controllers"
+	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/conversion"
 	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/listener"
 	"github.com/kubernetes-sigs/multi-tenancy/incubator/virtualcluster/pkg/syncer/manager"
-	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
 )
 
 const (
-	KubeconfigAdmin       = "admin-kubeconfig"
-	TLSTimeoutRetryPeriod = 1 * time.Second
+	KubeconfigAdmin = "admin-kubeconfig"
 )
 
 type Syncer struct {
+	sync.Mutex
 	secretClient      v1core.SecretsGetter
 	controllerManager *manager.ControllerManager
+	// lister that can list virtual clusters from a shared cache
+	lister vclisters.VirtualclusterLister
+	// returns true when the namespace cache is ready
+	virtualClusterSynced cache.InformerSynced
+	// virtual cluster that have been queued up for processing by workers
+	queue   workqueue.RateLimitingInterface
+	workers int
```

Comment:
For this controller, I think we should force to do sequential reconcile as of now.

type MultiClusterController struct {
	// name is used to uniquely identify a Controller in tracing, logging and monitoring.  Name is required.
	name string

	// objectType is the type of object to watch.  e.g. &v1.Pod{}
	objectType runtime.Object

	// clusters is the internal cluster set this controller watches.
	clusters map[Cluster]struct{}

	Options
}
is not thread safe.